File: app/api/papers.py
from pathlib import Path
from typing import Literal

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile, status
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.job import JobOut
from app.schemas.paper import PaperFromUrl, PaperListResponse, PaperOut, PaperUpdate
from app.services import jobs as jobs_service
from app.services import papers as papers_service
from app.services.documents import (
    document_chunk_payload,
    document_path,
    load_document_json,
    paper_dir,
)

router = APIRouter(prefix="/papers", tags=["papers"])


@router.get("", response_model=PaperListResponse)
def list_papers(
    folder_id: str | None = Query(None),
    view: Literal["all", "unfiled", "processing", "recent", "trash"] = Query("all"),
    query: str = Query("", max_length=200),
    sort: Literal["updated", "created", "title", "opened"] = Query("updated"),
    db: Session = Depends(get_db),
) -> PaperListResponse:
    items = papers_service.list_papers(
        db, folder_id=folder_id, view=view, query=query, sort=sort
    )
    return PaperListResponse(items=items, total=len(items))


@router.post("", response_model=PaperOut, status_code=201)
async def upload_paper(
    file: UploadFile = File(...),
    folder_id: str | None = Form(None),
    db: Session = Depends(get_db),
) -> PaperOut:
    paper = await papers_service.create_paper_from_upload(db, file, folder_id)
    return PaperOut.model_validate(paper)


@router.post("/from-url", response_model=PaperOut, status_code=201)
def import_paper_from_url(body: PaperFromUrl, db: Session = Depends(get_db)) -> PaperOut:
    """下载 arXiv PDF 到本地 uploads，再入队解析。"""
    paper = papers_service.create_paper_from_arxiv(db, body.url, body.folder_id)
    return PaperOut.model_validate(paper)


@router.get("/{paper_id}", response_model=PaperOut)
def get_paper(paper_id: str, db: Session = Depends(get_db)) -> PaperOut:
    return PaperOut.model_validate(papers_service.get_paper(db, paper_id))


@router.patch("/{paper_id}", response_model=PaperOut)
def update_paper(paper_id: str, body: PaperUpdate, db: Session = Depends(get_db)) -> PaperOut:
    fields = body.model_fields_set
    paper = papers_service.update_paper(
        db,
        paper_id,
        title=body.title,
        set_title="title" in fields,
        folder_id=body.folder_id,
        set_folder="folder_id" in fields,
    )
    return PaperOut.model_validate(paper)


@router.delete("/{paper_id}", status_code=204)
def delete_paper(paper_id: str, db: Session = Depends(get_db)) -> None:
    papers_service.delete_paper(db, paper_id)


@router.post("/{paper_id}/restore", response_model=PaperOut)
def restore_paper(paper_id: str, db: Session = Depends(get_db)) -> PaperOut:
    return PaperOut.model_validate(papers_service.restore_paper(db, paper_id))


@router.post("/{paper_id}/opened", response_model=PaperOut)
def mark_paper_opened(paper_id: str, db: Session = Depends(get_db)) -> PaperOut:
    return PaperOut.model_validate(papers_service.mark_opened(db, paper_id))


@router.post("/{paper_id}/metadata", response_model=PaperOut)
def refresh_paper_metadata(paper_id: str, db: Session = Depends(get_db)) -> PaperOut:
    return PaperOut.model_validate(papers_service.refresh_paper_metadata(db, paper_id))


@router.delete("/{paper_id}/permanent", status_code=204)
def permanently_delete_paper(paper_id: str, db: Session = Depends(get_db)) -> None:
    papers_service.permanently_delete_paper(db, paper_id)


@router.get("/{paper_id}/file")
def download_paper_file(paper_id: str, db: Session = Depends(get_db)) -> FileResponse:
    path: Path = papers_service.resolve_paper_file(db, paper_id)
    paper = papers_service.get_paper(db, paper_id)
    return FileResponse(
        path,
        media_type="application/pdf",
        filename=paper.filename,
        content_disposition_type="inline",
    )


@router.get("/{paper_id}/document")
def get_document(paper_id: str, db: Session = Depends(get_db)) -> FileResponse:
    papers_service.get_paper(db, paper_id)
    path = document_path(paper_id)
    if not path.exists():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="文档尚未解析完成")
    return FileResponse(path, media_type="application/json")


@router.get("/{paper_id}/document/chunk")
def get_document_chunk(
    paper_id: str,
    start_page: int = Query(1, ge=1),
    page_limit: int = Query(8, ge=1, le=32),
    include_manifest: bool = Query(False),
    db: Session = Depends(get_db),
) -> dict:
    papers_service.get_paper(db, paper_id)
    document = load_document_json(paper_id)
    if document is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="文档尚未解析完成")
    return document_chunk_payload(
        document,
        start_page=start_page,
        page_limit=page_limit,
        include_manifest=include_manifest,
    )


@router.post("/{paper_id}/parse", response_model=JobOut, status_code=202)
def retry_parse(paper_id: str, db: Session = Depends(get_db)) -> JobOut:
    job = jobs_service.enqueue_parse_job(db, paper_id, reset_document=True)
    return JobOut.model_validate(job)


@router.get("/{paper_id}/jobs/latest", response_model=JobOut)
def latest_job(paper_id: str, db: Session = Depends(get_db)) -> JobOut:
    papers_service.get_paper(db, paper_id)
    job = jobs_service.latest_parse_job(db, paper_id)
    if job is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="暂无解析任务")
    return JobOut.model_validate(job)


@router.get("/{paper_id}/assets/{asset_path:path}")
def get_asset(paper_id: str, asset_path: str, db: Session = Depends(get_db)) -> FileResponse:
    papers_service.get_paper(db, paper_id)
    base = paper_dir(paper_id).resolve()
    target = (base / asset_path).resolve()
    if not str(target).startswith(str(base)) or not target.exists() or not target.is_file():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="资源不存在")
    return FileResponse(target)

File: frontend/src/api/papers.ts
import type {
  Folder,
  FolderListResponse,
  Paper,
  PaperListResponse,
  PaperQuery,
} from '@/types'

async function parseError(res: Response): Promise<string> {
  try {
    const data = await res.json()
    if (typeof data?.detail === 'string') return data.detail
    if (data?.detail?.message) return data.detail.message
    if (Array.isArray(data?.detail)) {
      return data.detail.map((d: { msg?: string }) => d.msg ?? JSON.stringify(d)).join('; ')
    }
    return JSON.stringify(data)
  } catch {
    return res.statusText || `请求失败 (${res.status})`
  }
}

export async function fetchPapers(options: PaperQuery = {}): Promise<PaperListResponse> {
  const query = new URLSearchParams()
  if (options.folderId) query.set('folder_id', options.folderId)
  if (options.view) query.set('view', options.view)
  if (options.query?.trim()) query.set('query', options.query.trim())
  if (options.sort) query.set('sort', options.sort)
  const suffix = query.size ? `?${query}` : ''
  const res = await fetch(`/api/papers${suffix}`)
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function uploadPaper(file: File, folderId?: string | null): Promise<Paper> {
  const form = new FormData()
  form.append('file', file)
  if (folderId) form.append('folder_id', folderId)
  const res = await fetch('/api/papers', { method: 'POST', body: form })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

/** 从 arXiv 链接或裸 ID 导入：服务端本地下载 PDF 后入队解析 */
export async function importPaperFromUrl(url: string, folderId?: string | null): Promise<Paper> {
  const res = await fetch('/api/papers/from-url', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ url, folder_id: folderId || null }),
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function fetchPaper(id: string): Promise<Paper> {
  const res = await fetch(`/api/papers/${id}`)
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function renamePaper(id: string, title: string): Promise<Paper> {
  const res = await fetch(`/api/papers/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title }),
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function movePaper(id: string, folderId: string | null): Promise<Paper> {
  const res = await fetch(`/api/papers/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ folder_id: folderId }),
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function deletePaper(id: string): Promise<void> {
  const res = await fetch(`/api/papers/${id}`, { method: 'DELETE' })
  if (!res.ok) throw new Error(await parseError(res))
}

export async function restorePaper(id: string): Promise<Paper> {
  const res = await fetch(`/api/papers/${id}/restore`, { method: 'POST' })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function permanentlyDeletePaper(id: string): Promise<void> {
  const res = await fetch(`/api/papers/${id}/permanent`, { method: 'DELETE' })
  if (!res.ok) throw new Error(await parseError(res))
}

export async function markPaperOpened(id: string): Promise<void> {
  const res = await fetch(`/api/papers/${id}/opened`, { method: 'POST' })
  if (!res.ok) throw new Error(await parseError(res))
}

export async function fetchFolders(): Promise<FolderListResponse> {
  const res = await fetch('/api/folders')
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function createFolder(name: string, parentId: string | null): Promise<Folder> {
  const res = await fetch('/api/folders', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name, parent_id: parentId }),
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function updateFolder(
  id: string,
  changes: { name?: string; parent_id?: string | null; sort_order?: number },
): Promise<Folder> {
  const res = await fetch(`/api/folders/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(changes),
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function deleteFolder(id: string): Promise<void> {
  const res = await fetch(`/api/folders/${id}`, { method: 'DELETE' })
  if (!res.ok) throw new Error(await parseError(res))
}

export function paperFileUrl(id: string): string {
  return `/api/papers/${id}/file`
}

export async function fetchDocument(id: string): Promise<import('@/types/document').DocumentModel> {
  const res = await fetch(`/api/papers/${id}/document`, { cache: 'no-store' })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function fetchDocumentChunk(
  id: string,
  startPage: number,
  pageLimit: number,
  options?: { includeManifest?: boolean; signal?: AbortSignal },
): Promise<import('@/types/document').DocumentChunk> {
  const query = new URLSearchParams({
    start_page: String(startPage),
    page_limit: String(pageLimit),
    include_manifest: String(!!options?.includeManifest),
  })
  const res = await fetch(`/api/papers/${id}/document/chunk?${query}`, {
    cache: 'no-store',
    signal: options?.signal,
  })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export async function retryParse(id: string): Promise<{ id: string; status: string }> {
  const res = await fetch(`/api/papers/${id}/parse`, { method: 'POST' })
  if (!res.ok) throw new Error(await parseError(res))
  return res.json()
}

export function paperAssetUrl(paperId: string, relativePath: string): string {
  return `/api/papers/${paperId}/assets/${relativePath.replace(/^\//, '')}`
}
